### PR TITLE
Model layer meta query

### LINKED ISSUE #287: orderby meta value?
I have a query that looks like this:

```
		 $recent = array(
					   'post_type' => 'post',
					   'paged' => ( get_query_var('paged') ? get_query_var('paged') : 1),
					   'posts_per_page' => '12',
						'orderby' => 'meta_value_num',
						'meta_key' => 'ratings',
						'order' => 'DESC',
						'cat' => $end
						
					   );
		$the_query = new WP_Query( $recent );
```

Is it possible to orderby meta key/value ? currently I am only seeing orderby options is limited to : Date, title, author, parent, slug, in.

Is it possible to extend the orderby options with custom functions of some sort?

Thanks

### LINKED ISSUE #729: Post order returned from cursors is not stable
When paginating data using Apollo I randomly get same items from different cursor pages

image

I cannot isolate this to a test because this happens randomly. I first noticed this while implementing meta value based ordering in #721 but today I noticed it happening without meta values too so I'm opening this new issue.

I'm not using custom ordering here but I'm using wp-graphql-tax-query based filtering.

Luckily my refactoring in #721 fixes this issue too. It changes the generated SQL query from

```sql
SELECT   wp_posts.ID FROM wp_posts  LEFT JOIN wp_term_relationships ON (wp_posts.ID = wp_term_relationships.object_id) WHERE 1=1  AND wp_posts.post_parent = 0  AND ( 
  ( 
    wp_term_relationships.term_taxonomy_id IN (23)
  )
) AND wp_posts.post_type = 'apartment' AND ((wp_posts.post_status = 'publish')) AND wp_posts.post_date <= '2019-03-18 19:04:34' AND wp_posts.ID != 481 GROUP BY wp_posts.ID ORDER BY wp_posts.post_date DESC LIMIT 0, 6
```

to

```sql
SELECT   wp_posts.ID FROM wp_posts  LEFT JOIN wp_term_relationships ON (wp_posts.ID = wp_term_relationships.object_id) WHERE 1=1  AND wp_posts.post_parent = 0  AND ( 
  ( 
    wp_term_relationships.term_taxonomy_id IN (23)
  )
) AND wp_posts.post_type = 'apartment' AND ((wp_posts.post_status = 'publish')) AND wp_posts.ID < 461 GROUP BY wp_posts.ID ORDER BY wp_posts.post_date DESC, wp_posts.ID DESC  LIMIT 0, 6
```

The post_date AND  `AND wp_posts.post_date <= '2019-03-18 19:04:30'` is gone and `, wp_posts.ID DESC` is added to the `GROUP BY` clause.

100% sure this is the best way to do but seems to fix my app. I can now get all values using cursors without duplicates.

### PR DESCRIPTION
**Pull request by epeli**
Same as #721 but targeting the model layer branch.

Closes #730
Closes #721
Fixes #729
Fixes #287

Unfortunately not all tests pass:

- `1) PostListWithDifferentStatusCept: Get posts with different post_status`
- `tests/wpunit/PostObjectConnectionQueriesTest.php:testFirstPost`

I wasn't able figure out yet why they fail. Keeping this as a draft now.

### CODE REVIEW

**Comment by epeli** [comment]
lol I don't known PHP

https://github.com/wp-graphql/wp-graphql/pull/744/commits/74dbe72ef768b22b0f1e3ed55ae927d67a0c9bfa

... `testFirstPost` passes now.

**Comment by jasonbahl** [comment]
@epeli you want to change this away from being a draft PR?

**Comment by jasonbahl** [comment]
@epeli ok, so the functional test was failing because of some Model Layer changes. 

The test was making a non-authenticated request and then querying for draft and published posts

Draft posts now require authentication to get back, so I adjusted the test to appropriately assert that if we query for content that's private and don't authenticate, we should only get the public data in response. 

I'm walking through the rest of the PR now, and will push any changes back up here if I have any

😄

File: src/Data/CursorBuilder.php [hunk 1]
```diff
@@ -17,30 +17,30 @@ class CursorBuilder {
 	 */
 	public $compare = null;
 
-	public function __construct( $compare = '>') {
```
Comment (epeli) [inline src/Data/CursorBuilder.php]: I miss Prettier already 😬

**Comment by jasonbahl** [comment]
Ya. There’s some things we might be able to do with GitHub actions, etc to handle formatting consistency automatically. 🤔

**Comment by jasonbahl** [comment]
@epeli I made a few changes. . .a bit of formatting (as noted above), and moving things into their own Cursor directory, as it looks like we'll add more of these Cursor classes soon.

**Comment by epeli** [comment]
Wohoo! 🎉

**Pull request merged**